File: servlet/spring-boot/java/oauth2/authorization-server/src/integTest/java/example/OAuth2AuthorizationServerApplicationITests.java
/*
 * Copyright 2021 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package example;

import java.util.Map;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Test;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.HttpHeaders;
import org.springframework.mock.web.MockHttpServletRequest;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;
import org.springframework.test.web.servlet.request.RequestPostProcessor;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

/**
 * Integration tests for {@link OAuth2AuthorizationServerApplication}.
 *
 * @author Steve Riesenberg
 */
@SpringBootTest
@AutoConfigureMockMvc
@ActiveProfiles("test")
public class OAuth2AuthorizationServerApplicationITests {

	private static final String CLIENT_ID = "messaging-client";

	private static final String CLIENT_SECRET = "secret";

	private final ObjectMapper objectMapper = new ObjectMapper();

	@Autowired
	private MockMvc mockMvc;

	@Test
	void performTokenRequestWhenValidClientCredentialsThenOk() throws Exception {
		// @formatter:off
		this.mockMvc.perform(post("/oauth2/token")
				.param("grant_type", "client_credentials")
				.param("scope", "message:read")
				.with(basicAuth(CLIENT_ID, CLIENT_SECRET)))
				.andExpect(status().isOk())
				.andExpect(jsonPath("$.access_token").isString())
				.andExpect(jsonPath("$.expires_in").isNumber())
				.andExpect(jsonPath("$.scope").value("message:read"))
				.andExpect(jsonPath("$.token_type").value("Bearer"));
		// @formatter:on
	}

	@Test
	void performTokenRequestWhenMissingScopeThenOk() throws Exception {
		// @formatter:off
		this.mockMvc.perform(post("/oauth2/token")
				.param("grant_type", "client_credentials")
				.with(basicAuth(CLIENT_ID, CLIENT_SECRET)))
				.andExpect(status().isOk())
				.andExpect(jsonPath("$.access_token").isString())
				.andExpect(jsonPath("$.expires_in").isNumber())
				.andExpect(jsonPath("$.scope").value("message:read message:write"))
				.andExpect(jsonPath("$.token_type").value("Bearer"));
		// @formatter:on
	}

	@Test
	void performTokenRequestWhenInvalidClientCredentialsThenUnauthorized() throws Exception {
		// @formatter:off
		this.mockMvc.perform(post("/oauth2/token")
				.param("grant_type", "client_credentials")
				.param("scope", "message:read")
				.with(basicAuth("bad", "password")))
				.andExpect(status().isUnauthorized())
				.andExpect(jsonPath("$.error").value("invalid_client"));
		// @formatter:on
	}

	@Test
	void performTokenRequestWhenMissingGrantTypeThenUnauthorized() throws Exception {
		// @formatter:off
		this.mockMvc.perform(post("/oauth2/token")
				.with(basicAuth("bad", "password")))
				.andExpect(status().isUnauthorized())
				.andExpect(jsonPath("$.error").value("invalid_client"));
		// @formatter:on
	}

	@Test
	void performTokenRequestWhenGrantTypeNotRegisteredThenBadRequest() throws Exception {
		// @formatter:off
		this.mockMvc.perform(post("/oauth2/token")
				.param("grant_type", "client_credentials")
				.with(basicAuth("login-client", "openid-connect")))
				.andExpect(status().isBadRequest())
				.andExpect(jsonPath("$.error").value("unauthorized_client"));
		// @formatter:on
	}

	@Test
	void performIntrospectionRequestWhenValidTokenThenOk() throws Exception {
		// @formatter:off
		this.mockMvc.perform(post("/oauth2/introspect")
				.param("token", getAccessToken())
				.with(basicAuth(CLIENT_ID, CLIENT_SECRET)))
				.andExpect(status().isOk())
				.andExpect(jsonPath("$.active").value("true"))
				.andExpect(jsonPath("$.aud[0]").value(CLIENT_ID))
				.andExpect(jsonPath("$.client_id").value(CLIENT_ID))
				.andExpect(jsonPath("$.exp").isNumber())
				.andExpect(jsonPath("$.iat").isNumber())
				.andExpect(jsonPath("$.iss").value("http://localhost:9000"))
				.andExpect(jsonPath("$.nbf").isNumber())
				.andExpect(jsonPath("$.scope").value("message:read"))
				.andExpect(jsonPath("$.sub").value(CLIENT_ID))
				.andExpect(jsonPath("$.token_type").value("Bearer"));
		// @formatter:on
	}

	@Test
	void performIntrospectionRequestWhenInvalidCredentialsThenUnauthorized() throws Exception {
		// @formatter:off
		this.mockMvc.perform(post("/oauth2/introspect")
				.param("token", getAccessToken())
				.with(basicAuth("bad", "password")))
				.andExpect(status().isUnauthorized())
				.andExpect(jsonPath("$.error").value("invalid_client"));
		// @formatter:on
	}

	private String getAccessToken() throws Exception {
		// @formatter:off
		MvcResult mvcResult = this.mockMvc.perform(post("/oauth2/token")
				.param("grant_type", "client_credentials")
				.param("scope", "message:read")
				.with(basicAuth(CLIENT_ID, CLIENT_SECRET)))
				.andExpect(status().isOk())
				.andExpect(jsonPath("$.access_token").exists())
				.andReturn();
		// @formatter:on

		String tokenResponseJson = mvcResult.getResponse().getContentAsString();
		Map<String, Object> tokenResponse = this.objectMapper.readValue(tokenResponseJson, new TypeReference<>() {
		});

		return tokenResponse.get("access_token").toString();
	}

	private static BasicAuthenticationRequestPostProcessor basicAuth(String username, String password) {
		return new BasicAuthenticationRequestPostProcessor(username, password);
	}

	private static final class BasicAuthenticationRequestPostProcessor implements RequestPostProcessor {

		private final String username;

		private final String password;

		private BasicAuthenticationRequestPostProcessor(String username, String password) {
			this.username = username;
			this.password = password;
		}

		@Override
		public MockHttpServletRequest postProcessRequest(MockHttpServletRequest request) {
			HttpHeaders headers = new HttpHeaders();
			headers.setBasicAuth(this.username, this.password);
			request.addHeader("Authorization", headers.getFirst("Authorization"));
			return request;
		}

	}

}

File: servlet/spring-boot/kotlin/oauth2/authorization-server/src/integTest/kotlin/example/OAuth2AuthorizationServerApplicationITests.kt
/*
 * Copyright 2021 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package example

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.HttpHeaders
import org.springframework.mock.web.MockHttpServletRequest
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.request.RequestPostProcessor
import org.springframework.test.web.servlet.result.MockMvcResultMatchers


/**
 * Integration tests for [OAuth2AuthorizationServerApplication].
 *
 * @author Steve Riesenberg
 */
@SpringBootTest
@AutoConfigureMockMvc
@ActiveProfiles("test")
class OAuth2AuthorizationServerApplicationITests {
    private val objectMapper = ObjectMapper()

    @Autowired
    private val mockMvc: MockMvc? = null

    @Test
    fun performTokenRequestWhenValidClientCredentialsThenOk() {
        // @formatter:off
        mockMvc!!.perform(MockMvcRequestBuilders.post("/oauth2/token")
            .param("grant_type", "client_credentials")
            .param("scope", "message:read")
            .with(basicAuth(CLIENT_ID, CLIENT_SECRET)))
            .andExpect(MockMvcResultMatchers.status().isOk)
            .andExpect(MockMvcResultMatchers.jsonPath("$.access_token").isString)
            .andExpect(MockMvcResultMatchers.jsonPath("$.expires_in").isNumber)
            .andExpect(MockMvcResultMatchers.jsonPath("$.scope").value("message:read"))
            .andExpect(MockMvcResultMatchers.jsonPath("$.token_type").value("Bearer"))
        // @formatter:on
    }

    @Test
    fun performTokenRequestWhenMissingScopeThenOk() {
        // @formatter:off
        mockMvc!!.perform(MockMvcRequestBuilders.post("/oauth2/token")
            .param("grant_type", "client_credentials")
            .with(basicAuth(CLIENT_ID, CLIENT_SECRET)))
            .andExpect(MockMvcResultMatchers.status().isOk)
            .andExpect(MockMvcResultMatchers.jsonPath("$.access_token").isString)
            .andExpect(MockMvcResultMatchers.jsonPath("$.expires_in").isNumber)
            .andExpect(MockMvcResultMatchers.jsonPath("$.scope").value("message:read message:write"))
            .andExpect(MockMvcResultMatchers.jsonPath("$.token_type").value("Bearer"))
        // @formatter:on
    }

    @Test
    fun performTokenRequestWhenInvalidClientCredentialsThenUnauthorized() {
        // @formatter:off
        mockMvc!!.perform(MockMvcRequestBuilders.post("/oauth2/token")
            .param("grant_type", "client_credentials")
            .param("scope", "message:read")
            .with(basicAuth("bad", "password")))
            .andExpect(MockMvcResultMatchers.status().isUnauthorized)
            .andExpect(MockMvcResultMatchers.jsonPath("$.error").value("invalid_client"))
        // @formatter:on
    }

    @Test
    fun performTokenRequestWhenMissingGrantTypeThenUnauthorized() {
        // @formatter:off
        mockMvc!!.perform(MockMvcRequestBuilders.post("/oauth2/token")
            .with(basicAuth("bad", "password")))
            .andExpect(MockMvcResultMatchers.status().isUnauthorized)
            .andExpect(MockMvcResultMatchers.jsonPath("$.error").value("invalid_client"))
        // @formatter:on
    }

    @Test
    fun performTokenRequestWhenGrantTypeNotRegisteredThenBadRequest() {
        // @formatter:off
        mockMvc!!.perform(MockMvcRequestBuilders.post("/oauth2/token")
            .param("grant_type", "client_credentials")
            .with(basicAuth("login-client", "openid-connect")))
            .andExpect(MockMvcResultMatchers.status().isBadRequest)
            .andExpect(MockMvcResultMatchers.jsonPath("$.error").value("unauthorized_client"))
        // @formatter:on
    }

    @Test
    fun performIntrospectionRequestWhenValidTokenThenOk() {
        // @formatter:off
        mockMvc!!.perform(MockMvcRequestBuilders.post("/oauth2/introspect")
            .param("token", accessToken)
            .with(basicAuth(CLIENT_ID, CLIENT_SECRET)))
            .andExpect(MockMvcResultMatchers.status().isOk)
            .andExpect(MockMvcResultMatchers.jsonPath("$.active").value("true"))
            .andExpect(MockMvcResultMatchers.jsonPath("$.aud[0]").value(CLIENT_ID))
            .andExpect(MockMvcResultMatchers.jsonPath("$.client_id").value(CLIENT_ID))
            .andExpect(MockMvcResultMatchers.jsonPath("$.exp").isNumber)
            .andExpect(MockMvcResultMatchers.jsonPath("$.iat").isNumber)
            .andExpect(MockMvcResultMatchers.jsonPath("$.iss").value("http://localhost:9000"))
            .andExpect(MockMvcResultMatchers.jsonPath("$.nbf").isNumber)
            .andExpect(MockMvcResultMatchers.jsonPath("$.scope").value("message:read"))
            .andExpect(MockMvcResultMatchers.jsonPath("$.sub").value(CLIENT_ID))
            .andExpect(MockMvcResultMatchers.jsonPath("$.token_type").value("Bearer"))
        // @formatter:on
    }

    @Test
    fun performIntrospectionRequestWhenInvalidCredentialsThenUnauthorized() {
        // @formatter:off
        mockMvc!!.perform(MockMvcRequestBuilders.post("/oauth2/introspect")
            .param("token", accessToken)
            .with(basicAuth("bad", "password")))
            .andExpect(MockMvcResultMatchers.status().isUnauthorized)
            .andExpect(MockMvcResultMatchers.jsonPath("$.error").value("invalid_client"))
        // @formatter:on
    }

    // @formatter:off
    private val accessToken:
    // @formatter:on
            String
        get() {
            // @formatter:off
            val mvcResult = mockMvc!!.perform(MockMvcRequestBuilders.post("/oauth2/token")
                .param("grant_type", "client_credentials")
                .param("scope", "message:read")
                .with(basicAuth(CLIENT_ID, CLIENT_SECRET)))
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andExpect(MockMvcResultMatchers.jsonPath("$.access_token").exists())
                .andReturn()
            // @formatter:on
            val tokenResponseJson = mvcResult.response.contentAsString
            val tokenResponse: Map<String, Any> =
                objectMapper.readValue(tokenResponseJson, object : TypeReference<Map<String, Any>>() {})
            return tokenResponse["access_token"].toString()
        }

    private class BasicAuthenticationRequestPostProcessor constructor(
        private val username: String,
        private val password: String,
    ) :
        RequestPostProcessor {
        override fun postProcessRequest(request: MockHttpServletRequest): MockHttpServletRequest {
            val headers = HttpHeaders()
            headers.setBasicAuth(username, password)
            request.addHeader("Authorization", headers.getFirst("Authorization")!!)
            return request
        }
    }

    companion object {
        private const val CLIENT_ID = "messaging-client"
        private const val CLIENT_SECRET = "secret"
        private fun basicAuth(username: String, password: String): BasicAuthenticationRequestPostProcessor {
            return BasicAuthenticationRequestPostProcessor(username, password)
        }
    }
}
